File: Multiple Linear Regression.R
# Multiple Linear Regression (MLR)

# Importing the dataset
ds = read.csv('50_Startups.csv')
# ds = ds[, 2:3] # If needing to select specific indexes

# Encoding catagorical data
ds$State = factor(ds$State, levels = c('New York', 'California', 'Florda'), labels = c(1, 2, 3))


# Splitting the dataset into a training set and test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(ds$Profit, SplitRatio = 0.8)
trs = subset(ds, split == TRUE) # Training Set
tes = subset(ds, split == FALSE) # Test Set

# Fitting the MLR to the training 
lr = lm(formula = Profit ~ ., data = trs) # Equivalant to profit ~ R.D.Spend + Administration + Marketing.Spend + State

# Predicting the Test set results
y_pred = predict(lr, newdata = tes)

# Building the optimal model using Backwards Elimination
lr = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
        data = trs)
summary(lr)

lr = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(lr)

lr = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(lr)

lr = lm(formula = Profit ~ R.D.Spend,
               data = dataset)
summary(lr)

File: Multiple Linear Regression.py
# Multiple Linear Regression (MLR)

# Import the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
ds = pd.read_csv("50_Startups.csv")
X = ds.iloc[:, : -1].values # [Rows, Columns] : = take all
Y = ds.iloc[:, 4].values

# Encoding catagorical data
# Dependant variable
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
le_X = LabelEncoder()
X[:, 3] = le_X.fit_transform(X[:, 3])
# Create dummy variables
oe = OneHotEncoder(categorical_features = [3])
X = oe.fit_transform(X).toarray()

# Avoiding the dummy variable trap
X = X[:, 1:] # Remove the first column - X[0]


# Splitting the dataset into a training set and test set
from sklearn.cross_validation import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Fitting the MLR to the training set
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, Y_train)

# Predicting the Test set results
y_pred = lr.predict(X_test)


# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1) # Adding a column of 1's
# Optimal matrix of features (Significant level is 5%)
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
lr_ols = sm.OLS(endog = Y, exog = X_opt).fit()
lr_ols.summary()

# Remove [2]
X_opt = X[:, [0, 1, 3, 4, 5]]
lr_ols = sm.OLS(endog = Y, exog = X_opt).fit()
lr_ols.summary()

# Remove [1]
X_opt = X[:, [0, 3, 4, 5]]
lr_ols = sm.OLS(endog = Y, exog = X_opt).fit()
lr_ols.summary()

# Remove [4]
X_opt = X[:, [0, 3, 5]]
lr_ols = sm.OLS(endog = Y, exog = X_opt).fit()
lr_ols.summary()

# Remove [5]
X_opt = X[:, [0, 3]]
lr_ols = sm.OLS(endog = Y, exog = X_opt).fit()
lr_ols.summary()
